### ISSUE TITLE
20200619 is broken: can't open maildir

### ISSUE THREAD

**Issue by mgedmin** [neomutt/neomutt#2382]
## Expected Behaviour

I wish to read my incoming mail.  When I'm done with my inbox, I expect to be able to change to a different folder.

## Actual Behaviour

"(no mailbox)" in status line, I still see my inbox contents for some reason, although I can't open or move between messages ("no mailbox open" shows at the bottom).

## Steps to Reproduce

1. open neomutt
2. press `c`, select `=IN.vim`, press Enter

## How often does this happen?

- Always

## When did it start to happen?

- When I upgraded
  Which version did you use to use?  20200501

## NeoMutt Version

```
NeoMutt 20200619
Copyright (C) 1996-2016 Michael R. Elkins ir kiti.
NeoMutt ateina ABSOLIUČIAI BE JOKIOS GARANTIJOS; dėl smulkmenų paleisk 'neomutt -vv.'
NeoMutt yra laisva programinė įranga, ir tu gali laisvai ją platinti su tam
tikromis sąlygomis; rašyk 'neomutt -vv' dėl smulkmenų.

System: Linux 5.4.0-37-generic (x86_64)
ncurses: ncurses 6.2.20200212 (compiled with 6.2.20200212)
libidn: 1.33 (compiled with 1.33)
GPGME: 1.13.1-unknown
GnuTLS: 3.6.13
libnotmuch: 5.2.0
storage: tokyocabinet

Configure options: --prefix=/usr --sysconfdir=/etc --with-mailpath=/var/mail --mixmaster --gnutls --gpgme --gss --sasl --lua --notmuch --tokyocabinet --full-doc

Compilation CFLAGS: -g -O2 -std=c99 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include -I/usr/include/lua5.3 -DNCURSES_WIDECHAR -isystem /usr/include/mit-krb5

Įprastos parinktys:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Kompiliavimo parinktys:
  -autocrypt +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens 
  +getaddrinfo +gnutls +gpgme +gss +hcache -homespool +idn +inotify 
  -locales_hack +lua +meta +mixmaster +nls +notmuch -openssl +pgp +regex +sasl 
  +smime -sqlite +start_color +sun_attachment +typeahead 
MAILPATH="/var/mail"
MIXMASTER="mixmaster"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

Jei norite daugiau sužinoti apie NeoMutt, aplankykite https://neomutt.org
Jei radote NeoMutt klaidą, praneškite apie ją
    https://github.com/neomutt/neomutt/issues
arba paštu į <neomutt-devel@neomutt.org>
```

## Extra Info

* Operating System and its version: Ubuntu 20.04 LTS
* Were you using multiple copies of NeoMutt at once? no
* Were you using 'screen' or 'tmux'? no
* Is your email local (maildir) or remote (IMAP)? local (maildir)

**Comment by mgedmin** [comment]
git bisect blames commit 0dfdf7a863401242f1a6e85dfac278db95e77a1f.

**Comment by ajakk** [comment]
I'm having a similar failure and `git bisect` blames the same commit.

**Comment by mschiff** [comment]
Same problem here. After some debugging I was blaming `folder-hook` somehow:
I have a multi-account setup and I am using folder-hooks to switch between accounts.

This used to work like this:
```
folder-hook '~/Mail/private/INBOX'      'source ~/.mutt/accounts/private'
folder-hook '~/Mail/work/INBOX'        'source ~/.mutt/accounts/work'

macro index <F3> "c ~/Mail/private/INBOX\r" "Go to private inbox"
macro index <F4> "c ~/Mail/work/INBOX\r" "Go to work inbox"
```
Its all local Maildir folders and I am using mbsync to Sync with my IMAP-Folders.

If I set an account on startup by using something like `neomutt -e "source ~/.mutt/accounts/private"` it still works in this version until I try to switch the account using the above macro. When I hit F3 or F4 than trouble begins and I see "no mailbox" and stuff like this.

The sourced account files will set things like:
```
set my_init=`/bin/echo -en "\033]0;Private\007" >/proc/$PPID/fd/1 `
set sendmail="msmtp -C $HOME/.mutt/msmtprc -a private"
set folder="~/Mail/private"
set postponed="+Drafts"
set spoolfile="+INBOX"
set record = "+Sent"
set header_cache="~/.cache/mutt/headers/private/"
unmailboxes = *
mailboxes = `find ~/Mail/private -type d -name cur -regextype posix-extended -not -regex '.*/(Learn|Junk)\.?.*' -printf '%h '`
```

I tried running with debugging and I am seeing "no such file or directory" messages with the failing mailboxes and printing the correct and existing paths in the error message.

**Comment by flatcap** [comment]
@mgedmin do you have any hooks at work?
I've struggled and failed to repeat the problem.

**Comment by mgedmin** [comment]
Yes, I have folder-hooks as well.

My mail lives in ~/Mail/Home/, which has a bunch of subdirectories (maildirs) named `INBOX`, `IN.vim`, `vim` etc.

My muttrc does
```
set folder=~/Mail/Home
source $HOME/.mutt/folders.home
folder-hook Home 'set folder=~/Mail/Home; source $HOME/.mutt/folders.home'
account-hook . 'unset tunnel'
```
and ~/.mutt/folders.home does
```
set folder=~/Mail/Home          # Where all your mailboxes are
set mbox=+mbox                  # Where my personal mail goes
set postponed=+postponed        # The name of postponed messages folder
set record=+sent-mail           # Where copies of sent mail are kept
set spoolfile=+INBOX            # The main incoming mailbox
unmailboxes *
mailboxes +IN.personal
mailboxes +INBOX
mailboxes /var/mail/mg
mailboxes +IN.root
...
mailboxes +IN.vim
...
# The rest of them, if any
mailboxes `echo $HOME/Mail/Home/IN.*`
##mailboxes +postponed  <-- https://github.com/neomutt/neomutt/issues/1342
unmailboxes +IN.SPAM
```

**Comment by flatcap** [comment]
I've pushed a branch [[devel/change-folder]](https://github.com/neomutt/neomutt/tree/debug/change-folder) with lots of extra logging.
If you have time, please can you post what's between `OP_MAIN_CHANGE_FOLDER start` and `OP_MAIN_CHANGE_FOLDER end`
Thanks

**Comment by mgedmin** [comment]
Where do the logs end up?  I don't see anything on screen, nor do I see any log files in $PWD or /tmp.

**Comment by flatcap** [comment]
sorry, `neomutt -d2` logs to `~/.neomuttdebug0`

**Comment by mgedmin** [comment]
```
[2020-06-23 16:10:30]<1> mutt_index_menu() Got op OP_MAIN_CHANGE_FOLDER (126)
[2020-06-23 16:10:30]<1> mutt_index_menu() OP_MAIN_CHANGE_FOLDER start
[2020-06-23 16:10:31]<1> mutt_index_menu() user entered >>=IN.github<<
[2020-06-23 16:10:31]<1> mx_path_probe() unable to stat =IN.github: Toks failas ar aplankas neegzistuoja (errno 2)
[2020-06-23 16:10:31]<1> mutt_index_menu() search for '=IN.github'
[2020-06-23 16:10:31]<1> mutt_index_menu() found: /home/mg/Mail/Home/IN.github ()
[2020-06-23 16:10:31]<1> change_folder_mailbox() change_folder_mailbox
[2020-06-23 16:10:31]<1> change_folder_mailbox() have Context
[2020-06-23 16:10:31]<1> change_folder_mailbox() close Context
[2020-06-23 16:10:31]<M> mx_mbox_close() Dėžutė yra nepakeista
[2020-06-23 16:10:31]<1> change_folder_mailbox() rc = 0
[2020-06-23 16:10:31]<1> change_folder_mailbox() save current: /home/mg/Mail/Home/IN.github
[2020-06-23 16:10:31]<1> change_folder_mailbox() pre-folder-hook
[2020-06-23 16:10:31]<1> mutt_folder_hook() folder-hook: '/home/mg/Mail/Home/IN.github', ''
[2020-06-23 16:10:31]<1> mutt_folder_hook() folder-hook '.' matches '/home/mg/Mail/Home/IN.github'
[2020-06-23 16:10:31]<1> mutt_folder_hook()     set index_format="%4C%?M?+& ?%Z %[%m-%d] %-15.15F %?M?<%4M>&(%4c)? %s"
[2020-06-23 16:10:31]<1> mutt_folder_hook() hook command succeeded
[2020-06-23 16:10:31]<1> mutt_folder_hook() folder-hook 'Home' matches '/home/mg/Mail/Home/IN.github'
[2020-06-23 16:10:31]<1> mutt_folder_hook()     set folder=~/Mail/Home; source $HOME/.mutt/folders.home
[2020-06-23 16:10:31]<2> source_rc() Reading configuration file '/home/mg/.mutt/folders.home'
[2020-06-23 16:10:31]<1> mutt_folder_hook() hook command succeeded
[2020-06-23 16:10:31]<1> change_folder_mailbox() post-folder-hook
[2020-06-23 16:10:31]<1> change_folder_mailbox() mailbox has GONE
[2020-06-23 16:10:31]<1> mutt_index_menu() OP_MAIN_CHANGE_FOLDER end
```

**Comment by flatcap** [comment]
Thanks.
I _think_ I might know how to trigger this, now.

**Comment by mgedmin** [comment]
/me looks at the `unmailboxes *` with suspicion.

**Comment by amenl** [comment]
Same.

Commenting 'unmailboxes *' let's me access mails again, but mailbox list in sidebar piles up mailboxes from other accessed accounts.

**Comment by flatcap** [comment]
@mgedmin Please can you give branch [[devel/change-folder]](https://github.com/neomutt/neomutt/tree/devel/change-folder) a try.

You too @amenl if you're technical and happy to build your own NeoMutt.

**Comment by amenl** [comment]
@mgedmin Standard arch package (community repository) here, see https://www.archlinux.org/packages/community/x86_64/neomutt/

**Comment by mgedmin** [comment]
The bug is fixed in devel/change-folder (commit 319102a49d3c89124812c3220ede3d54565fd767).
